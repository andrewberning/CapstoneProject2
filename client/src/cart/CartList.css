.CartList {
  margin:  20px 50px;
}

.CartList .cart-title-container {
  margin-bottom: 20px;
  text-align: center;
}

.CartList .cart-details-container {
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  gap: 20px;
}

.CartList .items-list {
  display: flex;
  flex-direction: column;
  padding: 0;
  gap: 15px;
}

.CartList .cart-items-container, .CartList .order-summary-container {
  flex: 1;
}

.CartList .cart-items-container h2 {
  text-align: center;
}

.CartList .order-summary-container {
  flex: 1;
  max-height: 300px; /* Adjust as needed */
  padding: 20px;
  border-radius: 5px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  background-color: #f8f9fa;
  position: sticky;
  top: 0; /* Keeps it fixed while scrolling */
}

.CartList .order-summary-total {
  margin: 0;
  font-size: 1.2rem;
}