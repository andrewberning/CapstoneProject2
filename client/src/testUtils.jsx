import React from "react";
import UserContext from "./auth/UserContext";
import { vi } from 'vitest';

const demoUser = {
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "testuser123@test.net",
};

// Mock functions
const mockSetCurrentUser = vi.fn();
const mockLogin = vi.fn();
const mockLogout = vi.fn();
const mockSignup = vi.fn();
const mockAddToCart = vi.fn();
const mockHasAddedToCart = vi.fn();

const UserProvider = ({ 
  children,
  currentUser = demoUser,
  setCurrentUser = mockSetCurrentUser,
  login = mockLogin,
  logout = mockLogout,
  signup = mockSignup,
  addToCart = mockAddToCart,
  hasAddedToCart = mockHasAddedToCart,
  cartItems = [],
  setCartItems = () => {},
  categories = [],
  infoLoaded = true,
  totalAmount = 0,
  totalItems = 0,
}) => (
  <UserContext.Provider value={{ 
    currentUser,
    setCurrentUser,
    login,
    logout,
    signup,
    addToCart,
    hasAddedToCart,
    cartItems,
    setCartItems,
    categories,
    infoLoaded,
    totalAmount,
    totalItems,
  }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider, mockSetCurrentUser, mockLogin, mockLogout, mockSignup, mockAddToCart, mockHasAddedToCart };
