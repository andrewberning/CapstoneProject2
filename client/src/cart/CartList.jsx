import { useContext, useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import CartCard from './CartCard';
import UserContext from '../auth/UserContext';
import ShoplyApi from '../api/api';
import "./CartList.css";

export default function CartList() {
  const { currentUser, totalAmount, totalItems } = useContext(UserContext);

  const [cartItems, setCartItems] = useState([]);

  useEffect(() => {
    async function fetchData() {
      const cartData = await ShoplyApi.getCartItems(currentUser.cartId);
      setCartItems(cartData);
    }

    fetchData();
  }, [])

  const handleRemove = async (itemId) => {
    try {
      await ShoplyApi.removeFromCart(itemId);
      setCartItems(cartItems.filter(item => item.id !== itemId));
    } catch (err) {
      console.error("Error removing item from cart", err);
    }
  }

  const handleQuantityChange = async (itemId, newQuantity) => {
    try {
      await ShoplyApi.updateCartItemQuantity(itemId, newQuantity);
      setCartItems(cartItems.map(item => 
        item.id === itemId ? { ...item, quantity: newQuantity } : item
      ));
    } catch (err) {
      console.error("Error updating item quantity", err);
    }
  };

  return (
    <div className="CartList">
      <div className="cart-title-container">
        <h2>{currentUser.firstName}'s Cart</h2>
      </div>
      
      <div className="cart-details-container">
        <div className='cart-items-container'>
          <h2>Cart Items</h2>
          {cartItems.length > 0 ? (
            <ul className="items-list">
              {cartItems.map(item => (
                <CartCard 
                  key={item.id}
                  id={item.id}
                  productId={item.product_id}
                  name={item.name}
                  desc={item.description}
                  price={item.price}
                  img={item.image_url}
                  quantity={item.quantity}
                  categoryId={item.category_id}
                  onRemove={handleRemove}
                  onQuantityChange={handleQuantityChange}
                />
              ))}
            </ul>
          ) : <h2>Your cart is empty</h2>
          }
        </div>
        <div className="order-summary-container">
          <div className="order-summary-title">
            <h4>Order Summary</h4>
          </div>
          <div className="order-summary-total">
            <p>Number of items: {totalItems}</p>
            <p>Subtotal: ${totalAmount.toFixed(2)}</p>
          </div>
          <Link to="/checkout" className="btn btn-primary mt-3">
            Proceed to Checkout
          </Link>
        </div>
      </div>
    </div>
  );
}