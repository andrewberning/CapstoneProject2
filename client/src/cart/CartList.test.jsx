import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import CartList from "./CartList";
import { UserProvider } from "../testUtils";

describe('CategoriesBanner component', () => {
  it('renders without crashing', () => {
    render(
      <MemoryRouter>
        <UserProvider >
          <CartList />
        </UserProvider>
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <CartList />
        </UserProvider>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});