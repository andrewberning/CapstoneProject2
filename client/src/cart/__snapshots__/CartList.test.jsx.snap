// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoriesBanner component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CartList"
  >
    <div
      class="cart-title-container"
    >
      <h2>
        's Cart
      </h2>
    </div>
    <div
      class="cart-details-container"
    >
      <div
        class="cart-items-container"
      >
        <h2>
          Cart Items
        </h2>
        <h2>
          Your cart is empty
        </h2>
      </div>
      <div
        class="order-summary-container"
      >
        <div
          class="order-summary-title"
        >
          <h4>
            Order Summary
          </h4>
        </div>
        <div
          class="order-summary-total"
        >
          <p>
            Number of items: 0
          </p>
          <p>
            Subtotal: $0.00
          </p>
        </div>
        <a
          class="btn btn-primary mt-3"
          href="/checkout"
        >
          Proceed to Checkout
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
